src/relax/transform/to_mixed_precision.cc

```diff
@@ -0,0 +1,479 @@
+  bool AllFp16Available(const Array<Expr>& args) {
+    // TODO(@bohan): implements a tuple visitor that accepts a leaf handler to replace this logic
+    for (const Expr& arg : args) {
+      // arg is either a var or a constant or a tuple
+      if (const auto* var_node = arg.as<VarNode>()) {
+        // arg is a var
+        auto it = realized_var_.find(GetRef<Var>(var_node));
+        ICHECK(it != realized_var_.end()) << "Var not found: " << arg;
+        Var realized = it->second;
+        NDType realized_sinfo = NDTypeFrom(realized);
+        NDType realized_fp16 = NDTypeFrom(realized, fp16_);
+        if (!NDTypeEqual()(realized_fp16, realized_sinfo)) {
+          return false;
+        }
+      } else if (const auto* tuple_node = arg.as<TupleNode>()) {
+        if (!AllFp16Available(tuple_node->fields)) {
+          return false;
+        }
+      }
+      // constant can always be casted, so we don't need to check it
+    }
+    return true;
+  }
+
+  void VisitBinding_(const VarBindingNode* binding, const CallNode* call_node) override {
+    Optional<Array<ObjectRef>> dtype_convert_info = GetDTypeConvertInfo(call_node, output_dtype_);
+    if (!dtype_convert_info.defined()) {
+      const auto* op = call_node->op.as<OpNode>();
+      if (op == nullptr || !wrap_param_op.same_as(GetRef<Op>(op))) {
+        ExprMutator::VisitBinding_(binding, call_node);
+        return;
+      }
+      // wrap_param
+      const auto* constant = call_node->args[0].as<ConstantNode>();
+      ICHECK(constant != nullptr) << "Invalid wrap_param: " << GetRef<Call>(call_node);
+      ReEmitBinding(binding, builder_->Normalize(GetRef<Constant>(constant)));
+      return;
+    }
+    // var: NDType = Call(Op, [args], attr)
+    Array<ObjectRef> res = dtype_convert_info.value();
+    MixedTypePolicy policy = static_cast<MixedTypePolicy>(Downcast<Integer>(res[0])->value);
+    Call adjusted_call = Downcast<Call>(res[1]);
+    ObjectPtr<CallNode> new_call = make_object<CallNode>(*adjusted_call.get());
+    new_call->struct_info_ = NullOpt;
+    DataType to;
+    if (policy == MIXED_PRECISION_ALWAYS) {
+      to = fp16_;
+    } else if (policy == MIXED_PRECISION_FOLLOW) {
+      to = AllFp16Available(call_node->args) ? fp16_ : fp32_;
+    } else if (policy == MIXED_PRECISION_NEVER) {
+      to = fp32_;
+    } else {
+      LOG(FATAL) << "Unsupported MixedTypePolicy: " << policy;
+    }
+    Array<Expr> new_args = RewriteArgs(call_node->args, to);
+    new_call->args = std::move(new_args);
+    ReEmitBinding(binding, builder_->Normalize(Call(new_call)));
+  }
+
+  void VisitBinding_(const VarBindingNode* binding, const TupleNode* val) override {
+    // var = (v0, v1, ...)
+    auto it = var_dtype_map_->find(binding->var);
+    if (it != var_dtype_map_->end()) {
+      ReEmitBinding(binding, builder_->Normalize(RewriteExpr(GetRef<Tuple>(val), it->second)));
+      return;
+    }
+    // var is not NTensor or var is not used
```

Comment:
I guess “NTensor” means “nested Tensor,” but here’s the only appearance of “NTensor” so we probably need some elaboration if we want to use this word. Or we can change the word here.